json_encoder.go

```diff
@@ -158,37 +177,37 @@ func (enc *jsonEncoder) AddFields(fields []Field) {
 // the encoder's accumulated fields. It doesn't modify or lock the encoder's
 // underlying byte slice. It's safe to call from multiple goroutines, but it's
 // not safe to call WriteEntry while adding fields.
-func (enc *jsonEncoder) WriteEntry(sink io.Writer, e *Entry) error {
+func (enc *jsonEncoder) WriteEntry(sink io.Writer, msg string, lvl Level, t time.Time) error {
 	if sink == nil {
 		return errNilSink
 	}
-	if e == nil {
-		return errNilEntry
-	}
-	// Grab an encoder from the pool so that we can re-use the underlying
-	// buffer.
-	final := newJSONEncoder()
-	defer final.Free()
-
-	final.bytes = append(final.bytes, `{"msg":"`...)
-	final.safeAddString(e.Message)
-	final.bytes = append(final.bytes, `","level":"`...)
-	final.bytes = append(final.bytes, e.Level.String()...)
-	final.bytes = append(final.bytes, `","ts":`...)
-	final.bytes = strconv.AppendFloat(final.bytes, timeToSeconds(e.Time), 'f', -1, 64)
+
+	final := enc.Clone().(*jsonEncoder)
+
+	// TODO: When we flatten the serialized messages, remove the extra truncate
+	// and copy below. This will change the shape of the serialized message, which
+	// will require updating many tests.
+	final.truncate()
+	final.bytes = append(final.bytes, '{')
+	final.AddFields([]Field{
+		final.messageF(msg),
+		final.levelF(lvl),
+		final.timeF(t),
+	})
 	final.bytes = append(final.bytes, `,"fields":{`...)
-	if e.enc != nil {
-		final.bytes = append(final.bytes, e.enc.(*jsonEncoder).bytes...)
-	}
+	final.bytes = append(final.bytes, enc.bytes...)
 	final.bytes = append(final.bytes, "}}\n"...)
 
 	n, err := sink.Write(final.bytes)
 	if err != nil {
+		final.Free()
```

Comment:
consider breaking up this function into 2, one that writes out the entry given a `*jsonEncoder`, and the caller just gets the encoder + calls free afterwards. I'm not a fan of multiple return paths having to remember to `Free`